flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactory.scala

```diff
@@ -484,6 +493,9 @@ object FlinkTypeFactory {
       // CURSOR for UDTF case, whose type info will never be used, just a placeholder
       case CURSOR => new TypeInformationRawType[Nothing](new NothingTypeInfo)
 
+      case null if relDataType.isInstanceOf[RawRelDataType] =>
```

Comment:
nit: Wouldn't it be better to assign `SqlTypeName.OTHER` for the `RawRelDataType`? I am not sure about the consequences, but as far as I could look around in the codebase `RawRelDataType` would be the only type without `SqlTypeName`.

Maybe we could ask @danny0405 about his opinion here?

Reply:
I would suggest to use `SqlTypeName.ANY` which indicates the type comes from a Java object.

Reply:
This was actually a conscious decision. `SqlTypeName` is used (as the name suggests) for SQL types. All additional types don't need a type name. The usage of `SqlTypeName.ANY` in the past was actually a mistake. Because Calcite assumes that an `ANY` type does not need further parameters and `BasicSqlType` can be used a an instance class. The `ANY` type belongs to pure `Object.class` without any further parameters. I currently don't see a reason why we should come up with an arbitrary mapping to an enum value if this is actually not a SQL type.

Reply:
> Because Calcite assumes that an ANY type does not need further parameters
Calcite does not have such assumption. Maybe what you meant is a `BasicSqlType` with `SqlTypeName.ANY` does not need further parameters.

The `SqlTypeName` is a way to categorize these `RelDataType`, it does not indicates anything about the type instance itself.

I'm confused by "if this is actually not a SQL type", if we can use the `RAW(...)` in DDL or other DQLs, it is a SQL type.

Reply:
There is a difference between `RelDataType` and `AbstractSqlType`. Yes, we can use this type in DDL but this doesn't mean it is a valid SQL type. If you take a look at `SqlTypeName.ANY`, according to the definition there, it even takes precision and scale:
```
ANY(PrecScale.NO_NO | PrecScale.YES_NO | PrecScale.YES_YES
```

Reply:
If a type can be used in all kinds of SQL contexts directly, it is a SQL type, so the `RawRelDataType` should extend `AbstractSqlType` instead, as for the type name.

I'm not sure if a `RAW` type can has precision, either `SqlTypeName.ANY` or `SqlTypeName.OTHER` is fine for me, the precision constraint for type name is mainly used for sanity check when creating type instances in `RelDataTypeFactory`, Flink has it's own logic for creating this type, so this is not a problem.

Reply:
@danny0405 thanks for the hint around `SqlTypeName.OTHER`. This one fits better to the semantics of the RAW type than `SqlTypeName.ANY`.  The family ` SqlTypeFamily.ANY` includes all types which is not correct for the RAW type. But `SqlTypeName.OTHER` has no type family which is correct. I'm fine with using `SqlTypeName.OTHER`.